x/poolmanager/taker_fee.go

```diff
@@ -176,3 +176,110 @@ func CalcTakerFeeExactOut(tokenIn sdk.Coin, takerFee osmomath.Dec) (sdk.Coin, sd
 
 	return tokenInAfterAddTakerFee, takerFeeCoin
 }
+
+// TakerFeeSkim calculates the taker fee share for each denomination involved in a route and increases the accumulator for the respective denomination pair.
+// The function first sorts the denominations lexicographically and then checks for denomShareAgreement and alloyedAssetShareAgreement denoms.
+// DenomShareAgreement denoms represent a denom that has a taker fee share agreement with the Osmosis protocol, while alloyedAssetShareAgreement denoms represent a registered alloyed asset pool composed of one or more denoms with a denomShareAgreement.
+// If there are one or more denomShareAgreement denoms, the function calculates the percentage of the taker fees that should be skimmed off and increases the accumulator for the denomShareAgreement denom / taker fee denomination pair.
+// If there were no denomShareAgreement denoms but there are one or more alloyedAssetShareAgreement denoms, the function calculates the taker fee share for the alloyed asset for each underlying asset that has a taker fee share agreement.
+// The function returns an error if the total taker fee share percentage is greater than 1.
+//
+// Parameters:
+// - ctx: The context of the function call.
+// - denomsInvolvedInRoute: A slice of strings representing the denominations involved in the route.
+// - totalTakerFees: The total taker fees from the swap represented as sdk.Coins.
+//
+// Returns:
+// - An error if the total taker fee share percentage is greater than 1, or if there's an error in increasing the accumulator for the denomination pair.
+func (k Keeper) TakerFeeSkim(ctx sdk.Context, denomsInvolvedInRoute []string, totalTakerFees sdk.Coins) error {
+	// Sort the denoms involved in the route lexicographically.
+	osmoutils.SortSlice(denomsInvolvedInRoute)
+
+	// Retrieve the share agreements for denoms.
+	denomShareAgreements, alloyedAssetShareAgreements := k.getTakerFeeShareAgreements(denomsInvolvedInRoute)
+
+	// Process denom share agreements if any exist.
+	if len(denomShareAgreements) > 0 {
+		return k.processDenomShareAgreements(ctx, denomShareAgreements, totalTakerFees)
+	}
+
+	// Process alloyed asset share agreements if no denom share agreements exist.
+	if len(alloyedAssetShareAgreements) > 0 {
+		return k.processAlloyedAssetShareAgreements(ctx, alloyedAssetShareAgreements, totalTakerFees)
+	}
+
+	return nil
+}
+
+// getTakerFeeShareAgreements checks for individual denomShareAgreement and alloyedAssetShareAgreement denoms.
+func (k Keeper) getTakerFeeShareAgreements(denomsInvolvedInRoute []string) ([]types.TakerFeeShareAgreement, []types.TakerFeeShareAgreement) {
```

Comment:
Could you please document that if there is both `denomShareAgreements` and `alloyedAssetShareAgreements`, the alloyed is ignored?

Reply:
No longer the case after our call about changing this method, resolving.